### ISSUE TITLE
branches is not a valid map index

### ISSUE THREAD

**Issue by Vampire** [charleskorn/kaml#34]
This is the working JSON version:
**input file:**
```json
{
  "on": {
    "push": {
      "branches": [
        "master",
        "development"
      ]
    },
    "pull_request": null,
    "schedule": [
      {
        "cron": "0 0 * * FRI"
      }
    ]
  },
  "name": "Build and Test"
}
```
**DAO:**
```kt
@Serializable
data class GitHubWorkflow(
        val name: String? = null,
        val on: Map<String, OnDetails?>
) {
    sealed class OnDetails {
        @Serializable
        data class OnEventDetails(
                val details: Map<String, List<String>>?
        ) : OnDetails()

        @Serializable
        data class OnScheduleDetails(
                val details: List<Map<String, String>>
        ) : OnDetails()

        @Serializer(forClass = OnDetails::class)
        companion object : KSerializer<OnDetails?> {
            override fun deserialize(decoder: Decoder): OnDetails? {
                check(decoder is JsonInput) { "This class can only be loaded using JSON" }

                return when (val context = decoder.decodeJson()) {
                    is JsonArray -> OnScheduleDetails(decoder.json.parse(
                            MapSerializer(String.serializer(), String.serializer()).list, "$context")
                    )

                    is JsonObject -> OnEventDetails(decoder.json.parse(
                            MapSerializer(String.serializer(), String.serializer().list), "$context")
                    )

                    is JsonNull -> null

                    else -> error("Unexpected type for context: ${context::class}")
                }
            }

            override fun serialize(encoder: Encoder, value: OnDetails?) {
                when (value) {
                    is OnScheduleDetails -> encoder.encodeSerializableValue(
                            MapSerializer(String.serializer(), String.serializer()).list,
                            value.details
                    )

                    is OnEventDetails -> encoder.encodeSerializableValue(
                            MapSerializer(String.serializer(), String.serializer().list).nullable,
                            value.details
                    )
                }
            }
        }
    }
}
```

----

Now basically the same using YAML and kaml:
**input file:**
```yml
name: Build and Test

on:
    push:
        branches:
            - master
            - development
    pull_request:
    schedule:
        - cron: 0 0 * * FRI
```

**DAO:**
```kt
@Serializable
data class GitHubWorkflow(
        val name: String? = null,
        val on: Map<String, OnDetails?>
) {
    sealed class OnDetails {
        @Serializable
        data class OnEventDetails(
                val details: Map<String, List<String>>?
        ) : OnDetails()

        @Serializable
        data class OnScheduleDetails(
                val details: List<Map<String, String>>
        ) : OnDetails()

        @Serializer(forClass = OnDetails::class)
        companion object : KSerializer<OnDetails?> {
            override fun deserialize(decoder: Decoder): OnDetails? {
                check(decoder is YamlInput) { "This class can only be loaded using kaml" }

                val context = decoder.node
                check(context is YamlMap) { "Expected a YamlMap as current context" }

                return decoder.decodeStructure(descriptor) {
                    this as YamlInput

                    check(node is YamlMap) { "Expected a YamlMap as current context" }

                    return@decodeStructure if (context.getKey(node) == "schedule") {
                        OnScheduleDetails(decoder.decodeSerializableValue(
                                MapSerializer(String.serializer(), String.serializer()).list
                        ))
                    } else {
                        OnEventDetails(decoder.decodeSerializableValue(
                                MapSerializer(String.serializer(), String.serializer().list)
                        ))
                    }
                }
            }

            override fun serialize(encoder: Encoder, value: OnDetails?) {
                when (value) {
                    is OnScheduleDetails -> encoder.encodeSerializableValue(
                            MapSerializer(String.serializer(), String.serializer()).list,
                            value.details
                    )

                    is OnEventDetails -> encoder.encodeSerializableValue(
                            MapSerializer(String.serializer(), String.serializer().list).nullable,
                            value.details
                    )
                }
            }
        }
    }
}

private fun YamlMap.getKey(value: YamlNode) =
        entries
                .asSequence()
                .find { it.value == value }
                ?.key
                ?.let { it as? YamlScalar }
                ?.content
```

----

While the JSON version works properly and outputs `GitHubWorkflow(name=Build and Test, on={push=OnEventDetails(details={branches=[master, development]}), pull_request=null, schedule=OnScheduleDetails(details=[{cron=0 0 * * FRI}])}, env=null, defaults=null)`, the YAML version throws the following error:

```stacktrace
Caused by: java.lang.IllegalArgumentException: branches is not a valid map index
        at kotlinx.serialization.internal.MapLikeDescriptor.getElementIndex(CollectionDescriptors.kt:53)
        at com.charleskorn.kaml.YamlObjectInput.decodeElementIndex(YamlInput.kt:384)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.patch(CollectionSerializers.kt:34)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.deserialize(CollectionSerializers.kt:45)
        at kotlinx.serialization.Decoder$DefaultImpls.decodeSerializableValue(Decoding.kt:225)
        at kotlinx.serialization.builtins.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:16)
        at com.charleskorn.kaml.YamlInput.decodeSerializableValue(YamlInput.kt:113)
        at net.kautler.dao.GitHubWorkflow$OnDetails$Companion.deserialize(GitHubWorkflow.kt:78)
        at net.kautler.dao.GitHubWorkflow$OnDetails$Companion.deserialize(GitHubWorkflow.kt:61)
        at kotlinx.serialization.Decoder$DefaultImpls.decodeSerializableValue(Decoding.kt:225)
        at kotlinx.serialization.builtins.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:16)
        at com.charleskorn.kaml.YamlInput.decodeSerializableValue(YamlInput.kt:113)
        at kotlinx.serialization.internal.NullableSerializer.deserialize(NullableSerializer.kt:38)
        at kotlinx.serialization.Decoder$DefaultImpls.decodeSerializableValue(Decoding.kt:225)
        at kotlinx.serialization.builtins.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:16)
        at com.charleskorn.kaml.YamlInput.decodeSerializableValue(YamlInput.kt:113)
        at kotlinx.serialization.builtins.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:63)
        at kotlinx.serialization.internal.MapLikeSerializer.readElement(CollectionSerializers.kt:120)
        at kotlinx.serialization.internal.MapLikeSerializer.readElement(CollectionSerializers.kt:92)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.readElement$default(CollectionSerializers.kt:54)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.patch(CollectionSerializers.kt:36)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.deserialize(CollectionSerializers.kt:45)
        at kotlinx.serialization.Decoder$DefaultImpls.decodeSerializableValue(Decoding.kt:225)
        at kotlinx.serialization.builtins.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:16)
        at com.charleskorn.kaml.YamlInput.decodeSerializableValue(YamlInput.kt:113)
        at kotlinx.serialization.builtins.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:63)
        at net.kautler.dao.GitHubWorkflow$$serializer.deserialize(GitHubWorkflow.kt)
        at net.kautler.dao.GitHubWorkflow$$serializer.deserialize(GitHubWorkflow.kt:45)
        at kotlinx.serialization.Decoder$DefaultImpls.decodeSerializableValue(Decoding.kt:225)
        at kotlinx.serialization.builtins.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:16)
        at com.charleskorn.kaml.YamlInput.decodeSerializableValue(YamlInput.kt:113)
        at kotlinx.serialization.DecodingKt.decode(Decoding.kt:521)
        at com.charleskorn.kaml.Yaml.parse(Yaml.kt:43)
        ... 172 more
```

**Comment by Vampire** [comment]
Ah, this seems to be another manifestation of the different descriptor / custom serializer handling. If I add a manual descriptor `override val descriptor: SerialDescriptor = SerialDescriptor("net.kautler.dao.GitHubWorkflow.OnDetails", CONTEXTUAL)` it works.

**Comment by charleskorn** [comment]
OK, I'm going to close this as a duplicate of #30 then - feel free to reopen this if you feel that's incorrect.

**Comment by Vampire** [comment]
We will see if it is a duplicate or only similar in case you fix #30 :-D

But just for you to know, I cannot technically reopen an issue you closed unless you make me maintainer of the repository.
Only maintainer can re-open issues and the on that closed it.
So if I close it, I can reopen it.
But if you close it and say "re-open if needed", then this is unfortunately not possible on GitHub, we can just add more comments as long as the issue does not get locked.

**Comment by charleskorn** [comment]
Ah sorry - if #30 doesn't fix this, please just comment here and I'll reopen the issue.
